doc/geomview.1



Geomview(1)					      Geomview(1)


NAME
       Geomview - interactive geometry viewer

SYNOPSIS
       geomview [-opengl] [-b r g b] [-c commandfile]
	    [-wins #windows] [-nopanels]
	    [-wpos xmin ymin xsize ysize]
	    [-wpos xsize,ysize[@xmin,ymin]] [-e external-module-
	    name] [-M[cg][sp] pipename]
	    [-start external-module-name [arg ...] --]
	    [-run external-module-path [arg ...] --] [file ...]

DESCRIPTION
       Geomview	 is an interactive geometry viewer written at the
       Geometry Center.	 It displays the  objects  in  the  files
       given  on the command line and allows the user to view and
       manipulate them interactively.

       The present version (1.7) of geomview	runs  on  Silicon
       Graphics	 Irises,  and X Window System workstations.
       We are very interested in hearing about any  problems  you
       may  have  using	 it;  see below for information on how to
       contact us.

       In addition to providing interactive control over a  world
       of objects via the mouse and keyboard, geomview has an in-
       terpreted language of commands for controlling almost  ev-
       ery  aspect of its display.  It can execute files contain-
       ing statements in this language, and  it	 can  communicate
       with  other programs using the language.	 See geomview(5),
       e.g. the file man/cat5/geomview.5 in the geomview  distri-
       bution, for more details of the language.


TUTORIAL
       This  manual  page serves only as a (very) terse reference
       manual for geomview.  For a gentler  introduction  to  the
       program	and the format of the data files it can read, see
       "overview" and "oogltour" in the "doc"  directory  of  the
       geomview	 distribution,	or  better,  see the full manual:
       "geomview.tex" or "geomview.ps", also in the "doc"  direc-
       tory.   The  source  distribution also includes a tutorial
       for how to write external modules in the "src/bin/example"
       directory.


OPTIONS
       -opengl
	      Some  geomview  distributions  include two gvx exe-
	      cutables -- one which displays images using only	X
	      windows  graphics,  and another which uses Open GL.
	      The -opengl (or -OGL) option selects the latter, if
	      it  exists.  If used, this option must be the first
	      option given to geomview.



Geometry Center		December 10, 1996			1





Geomview(1)					      Geomview(1)


       -b     Set the window background color to the given r g	b
	      values.

       -c     Interpret the commands in commandfile, which may be
	      ``-'' for standard input.	 The command language  is
	      described in geomview(5).	 Commands may be supplied
	      literally, as  in	 ``-c  "(ui-panel  main	 off)"'';
	      since  they include parentheses, they must be quot-
	      ed.

       -wins  Causes geomview to initially display #windows  cam-
	      era windows.

       -wpos  Specifies	 the  initial  location	 and  size of the
	      first camera window.  With ``-'', you are	 prompted
	      for window placement.

       -M objname
	       Display	(possibly  dynamically changing) geometry
	      (or commands) sent via ``togeomview -g objname [ge-
	      omview-shell-command]'' or ``geomstuff objname file
	      ...''.  The data appears as a geomview object named
	      objname.

	      Actually	listens	 to  the  named	 pipe  ``/tmp/ge-
	      omview/objname''; you can achieve the  same  effect
	      with the shell commands:
		  mkdir /tmp/geomview;
		  mknod /tmp/geomview/objname p
	      (assuming	 the  directory	 and named pipe don't al-
	      ready exist), then executing the geomview command:
		  (geometry objname  < /tmp/geomview/objname)

       -M[cg][sp] pipename
	      The -M option accepts modifiers: a 'g'  suffix  ex-
	      pects geometry data (the default), while 'c' suffix
	      expects gcl commands.  A 'p' implies the connection
	      should  use a named pipe (the default on the Iris),
	      while 's' implies using a UNIX-domain  socket  (de-
	      fault  on	 the NeXT).  So -Mcs fred selects reading
	      commands from the UNIX-domain socket named /tmp/ge-
	      omview/fred.

       -nopanels
	      Start  up	 displaying no panels, only graphics win-
	      dows.  Panels may be invoked later  as  usual  with
	      the  "Px"	 keyboard  shortcuts  or "(ui-panel ...)"
	      command.

       -noinit
	      Read no initialization files.  By default, geomview
	      reads the system-wide ".geomview" file, followed by
	      those in $HOME/.geomview and ./.geomview.




Geometry Center		December 10, 1996			2





Geomview(1)					      Geomview(1)


       -e modulename
	      Start an external module; modulename  is	the  name
	      associated  with	the module, appearing in the main
	      panel's Applications browser,  as	 defined  by  the
	      emodule-define  command  (see  geomview(5)  for de-
	      tails).

       -start modulename arguments ... --
	      Like -e but allows you to pass arguments to the ex-
	      ternal  module.	"--" signals the end of the argu-
	      ment list; the "--" may be omitted if it	would  be
	      the last argument on the geomview command line.

       -run shell-command arguments ...
	      Like -start but takes the pathname of executable of
	      the external module instead of the module's name.


GEOMETRY FILE FORMATS
       The format of the files read by geomview is  described  in
       oogl(5);	  type	 "man	5   oogl",   or	  see	the  file
       man/cat5/oogl.5 in the geomview distribution, for details.

       Note  to	 users	of MinneView (the precursor to geomview):
       geomview can read MinneView files,  but	MinneView  cannot
       read all geomview files.


STARTUP FILES
       Immediately  upon  starting up geomview reads and executes
       the commands in the system-wide .geomview file in the "da-
       ta"  subdirectory  of  the  geomview  directory.	 Then, if
       there is a file named .geomview in the current  directory,
       it  executes  the commands in that file.	 If no in the us-
       er's home directory, and executes it if found. The startup
       file  of	 an  individual user overrides the systemwide de-
       faults, since later commands take precedence over  earlier
       ones.


EXTERNAL MODULES
       Geomview	 has the ability to interact via its command lan-
       guage with  other  programs,  called  "external	modules".
       Many such modules have been written and appear in the "Ap-
       plication" browser in the main geomview panel.  To  invoke
       a module you click the mouse on the module's entry in this
       browser.	 This starts the module and  adds  an  additional
       entry  to  the  browser, beginning with a number in square
       brackets as in ``[1] Crayola'', which represents the  run-
       ning  instance of that module.  You can terminate the mod-
       ule by clicking on the numbered entry.  Modules are  docu-
       mented  separately from geomview.  See the manual page for
       each module for details.




Geometry Center		December 10, 1996			3





Geomview(1)					      Geomview(1)


INSTALLING AN EXTERNAL MODULE
       Geomview looks for external modules in a special directory
       for  modules.   In  the geomview distribution tree this is
       the "bin/$MACHTYPE" subdirectory.  A  module  consists  of
       two  files:  the	 executable  program,  and a "module init
       file", which is a whose name is ".geomview-"  followed  by
       the  module name.  The module init file tells geomview how
       to run that program.  Be sure to	 always	 keep  these  two
       files  together;	 geomview  needs both of them in order to
       run the module.	To install a new module, simply	 put  the
       module's	 executable  file  and	its init file in your ge-
       omview's module directory.  The	next  time  you	 run  ge-
       omview, it will know about that module.

       Geomview	 can  actually looks for modules in a list of di-
       rectories; by default only the  "bin/$MACHTYPE"	directory
       is  on this list.  See the set-emodule-path command in ge-
       omview(5) for details.

       There is a tutorial for how to write external  modules  in
       the "src/bin/example" directory.


EXTERNAL MODULE INIT FILES
       An  external  module  init file is the file that tells ge-
       omview how to run that module.  Its  name  must	be  ".ge-
       omview-"	 followed  by  the name of the module, e.g. ".ge-
       omview-foo".   It should contain geomview commands;  typi-
       cally  it  will	contain	 a  single emodule-define command
       which enters the module into geomview's application brows-
       er:

	       (emodule-define "Foo" "foo")

       The  first string is the name that appears in the browser.
       The second string is the command to invoke the module.  It
       may  contain  arguments;	 in  fact  it can be an arbitrary
       shell command.



KEYBOARD SHORTCUTS
       Many geomview operations are available from the	keyboard.
       Hitting	the  "?"  button on the main panel, or typing "?"
       with the cursor in any window, causes geomview to print	a
       message on standard output listing all the keyboard short-
       cuts.

       Keyboard commands apply while cursor is in any graphics window and most
       control panels. Most commands allow one of the following selection prefixes
       (if none is provided the command applies to the current object):
	  g  world geom	   g#  #'th geom  g*  All geoms
	  c  current camera	c#  #'th camera	    c*	All cameras
       Many allow a numeric prefix:  if none they toggle or reset current value.



Geometry Center		December 10, 1996			4





Geomview(1)					      Geomview(1)


       Appearance:
	Draw:		   Shading:	  Other:
	 af  Faces	   0as Constant	   av  eVert normals: always face viewer
	 ae  Edges	   1as Flat  #aw  Line Width (pixels)
	 an  Normals	   2as Smooth	  #ac  edges Closer than faces(try 5-100)
	 ab  Bounding Boxes	3as Smooth, non-lighted	 al  Shade lines
	 aV  Vectors	   aT  allow transparency   at	Texture-mapping
	Color:		   aC  allow concave polygons
	 Cf Ce Cn Cb CB	  face/edge/normal/bbox/backgnd
       Motions:			      Viewing:
	 r rotate	 [ Leftmouse=X-Y plane,	    0vp Orthographic view
	 t translate	   Middle=Z axis,      1vp Perspective view
	 z zoom FOV	   Shift=slow motion,	     vd Draw other views' cameras
	 f fly		   in r/t modes.      ]	    #vv field of View
	 o orbit	   [Left=steer, Middle=speed ]	 #vn near clip distance
	 s scale		     #vf far clip distance
	 w/W recenter/all		   v+ add new camera
	 h/H halt/halt all		   vx cursor on/off
	 @  select center of motion (e.g. g3@)	     vb backfacing poly cull on/off
				     #vl focal length
	 L  Look At object		   v~ Software shading on/off
       show Panel:    Pm Pa Pl Po    main/appearance/lighting/obscure
		 Pt Pc PC Pf	tools/cameras/Commands/file-browser
		 Ps P-		saving/read commands from tty
       Lights:	ls le	   Show lights / Edit Lights
       Metric:	me mh ms   Euclidean Hyperbolic Spherical
       Model:	mv mp mc   Virtual Projective Conformal
       Other:
	 N normalization < Pf  load geom/command file
	  0N none	> Ps  save something to file	 ui  motion has inertia
	  1N each	TV NTSC mode toggle    uc  constrained (X/Y) motion
	  2N all		     uo	 motion in Own coord system
	 Rightmouse-doubleclick	 pick as current target object
	 Shift-Rightmouse	 pick interest (center) point
       Renderman:
	 RR send RIB output to <fileprefix>NNN.rib (default fileprefix == "geom")
	 RC Emulate lines using cylinders (default)
	 RP Emulate lines using polygons
	 Ra choose ASCII RIB format (default)
	 Rb choose BINARY RIB format
	 Rt choose Display token to specify .tiff file (default)
	 Rf choose Display token to specify framebuffer
	 Rs Simulate background color with Polygon (default)
	 Rx No background simulation - fully transparent (alpha) background


NOTES
       The "geomview" command is actually  a  shell  script  that
       sets  various  environment  variables  which tell geomview
       about your local setup, and then invokes the geomview exe-
       cutable program "gvx" (or "gvx.OGL").  Do not run "gvx" by
       itself; always invoke geomview with the	"geomview"  shell
       script.




Geometry Center		December 10, 1996			5





Geomview(1)					      Geomview(1)


SEE ALSO
       oogl(5) - OOGL geometric file formats and conventions
       geomview(5) - geomview command language reference

FILES
       data/.geomview  -  default initialization file in geomview
	    command language
       data/geom - sample data files

ENVIRONMENT
       The ``geomview'' shell script sets these internally by de-
       fault;  if you set them before invoking geomview, the val-
       ues you set will be used instead of the built-in defaults.
       GEOMROOT	 -  main directory under which geomview finds its
	    data files, modules, and gvx executable.
       GEOMVIEW_GVX - geomview executable
       GEOMVIEW_LOAD_PATH - colon-separated search path for  data
	    files
       GEOMVIEW_EMODULE_PATH  -	 colon-separated  search path for
	    external modules
       GEOMVIEW_SYSTEM_INITFILE - system-wide GCL  initialization
	    script
       GEOMDATA - top directory of the default data tree, used by
	    some modules


AUTHORS
	  Stuart Levy		   Tamara Munzner	  Mark Phillips
		    Celeste Fowler		Nathaniel Thurston
		     Daniel Krech		    Scott Wisdom
		     Daeron Meyer		   Timothy Rowley

	      The National Science and Technology Research Center for
	       Computation and Visualization of Geometric Structures
			       (The Geometry Center)
			      University of Minnesota

BUGS
       Sometimes core dumps on bad input files.

       Zoom and scale have no inertia.

       Virtual spherical mode doesn't work on VGXes.



















































Geometry Center		December 10, 1996			7




NEWS
1.8.2 August, 2006

   Bugfix release.

   * I/O with release <= 1.8.1 was just broken, they used internals of
     opaque data-types (e.g. the FILE handle of the stdio library)
   * Maple V8 interface
   * VRML2 interface
   * build-system fixes (ported to recent automake/autoconf combo)
   * standard GNU installation paths: PREFIX/bin, PREFIX/share/geomview,
     PREFIX/libexec/geomview etc.
   * variable argument fixes, was also broken in <= 1.8.1 releases.
   * Documentation can be regenerated with recent texinfo releases.
   * OpenGL is now optional
   * one big shared libgeomview library
   * fmemopen() by using temporary files (not using stdio internals)
   * I have resurrected a couple of old modules: NDview, clipboard,
     cplxview, crayola, flythrough, gvclock, hinge, labeler, ndutils.

   Otherwise no real changes.

   cH <Claus.Heine@Mathematik.Uni-Freiburg.DE>

1.8.1	March, 2001

   This is a bugfix release.

   * No segfaults at startup on linux.
   * Toggle buttons are visible again.
   * Binary I/O on linux works.

1.8.0   November 17, 2000

   The software in this release is almost the same as in the 1.7.*
   series. The main difference is that it has an improved configure
   script and Makefiles that allow for several new things.

   (a) VPATH builds now supported

   (b) Additional external modules can simply be dropped into place in
       the "src/bin" directory, and Geomview's configure script will
       notice their presence and configure/build them when you
       configure/build Geomview.  (This only works with module
       distributions that have been specifically packed up for this
       purpose; see http://www.geomview.org/download for more
       details.)

   (c) The binary Intel/Linux XFORMS distribution is no longer a part
       of the Geomview source distribution.  Users who want it can
       download it separately, unpack it into the Geomview source
       tree, and Geomview's configure script will notice it's there
       and use it.  (This works with the binary XFORMS distribution
       for any system, not just Intel/Linux.)

   (d) 'togeomview' program now included in the release (this program,
       used to facilitate communication between Geomview and other
       programs, got temporarily dropped in release 1.7.0, but now is
       back).

   (e) a few minor bug fixes; see ChangeLog for details


1.7.10  Nov 1, 2000

   Attempted to fix bug which caused Geomview to crash immediately
   upon startup on some newer GNU/Linux systems.  The bug has to do
   with Geomview's use of the 'fmemopen' system function.  This
   release attempts to autodetect which version of this function to
   use.  If it still crashes for you, pass the flag
   "--enable-fstropen-1" to "configure" and try again.  If you still
   have problems, let me [mbp@geomtech.com] know.

   Misc fixes to allow compilation on wider variety of platforms


1.7.9  Oct 20, 2000

   Minor change to allow compilation with Mesa-3.3 (related to how
   Mesa-3.3 declares prototypes for GL extention functions).  No
   change in functionality, but this version compiles without errors
   with Mesa-3.3, which previous versions didn't.

1.7.8  Oct 14, 2000     

   Fixed bug which prevented compressed textures from working on
   some Linux systems
        
   Fixed several compilation/configuration problems.  Now tested on:
       Linux:  RedHat 6.1, 7.0
	       Mandrake 7.0
	       Caldera-2.3
	       Mandrake-7.0
	       Slackware-7.0
       Solaris-2.6 (SunOS 5.6)
       AIX 4.3.2
       IRIX 6.3



1.7.7   August 22, 2000

   Added some of the external modules to the distribution, but haven't
   actually tested them.  Updated the documentation to reflect the
   fact that the geomview-users mailing list is now hosted at
   lists.sourceforge.net.

1.7.6   May 25, 2000

   This is the first release of Geomview to use GNU autoconf.  The
   autoconf script has been tested on Linux (RedHat 6.1), IRIX 6.3,
   and AIX 4.3.2.  It's untested on other platforms and may need
   tweaking to work on them.  Also, note that this release does not
   include the external modules.  If you want any of them, use an
   older release, or wait for a future one.


geomview.sh
#! /bin/sh

# geomview shell script driver

GEOMROOT=%GEOMROOT%
if test -x %GEOMVIEW_LTGVX%; then
    GEOMVIEW_GVX=%GEOMVIEW_LTGVX%
else
    GEOMVIEW_GVX=%GEOMVIEW_GVX%
fi

########################################################################
#
#		  DO NOT CHANGE ANYTHING BELOW HERE		       #
#
########################################################################

%LD_LIBRARY_PATH_SETTINGS%
export LD_LIBRARY_PATH

### Geomview and/or its modules use the following environment
### variables.  For each of these that does not already have a value,
### this shell script assigns a value based on the setting of
### geomview_dir above.  You can override these values if you want by
### setting them in your shell environment before invoking this
### script.
### 
### GEOMVIEW_DIR: pathname of geomview distribution directory
### GEOMVIEW_GVX: pathname of executable file "gvx"
### GEOMVIEW_LOAD_PATH: colon-separated list of directories to search
### 	for files
### GEOMVIEW_EMODULE_PATH: colon-separated list of directories to search
### 	for modules
### GEOMVIEW_SYSTEM_INITFILE: pathname of an initialization file to
### 	read upon startup
### GEOMDATA: [ used only by some modules; for backward compatibility ]
### 	pathname of the data directory

MACHTYPE=%MACHTYPE%

suf=

: ${GEOMDATA=%GEOMDATA%}

: ${GEOMVIEW_LOAD_PATH=.:${GEOMDATA}/geom:${GEOMDATA}}
: ${GEOMVIEW_EMODULE_PATH=%MODULESDIR%}
: ${GEOMVIEW_SYSTEM_INITFILE=${GEOMDATA}/.geomview}

export GEOMROOT GEOMVIEW_GVX GEOMVIEW_LOAD_PATH GEOMDATA \
        GEOMVIEW_EMODULE_PATH GEOMVIEW_SYSTEM_INITFILE

if test "`uname -o`" = Cygwin; then
    TCL_LIBRARY=/usr/share/tcl8.4
    TK_LIBRARY=/usr/share/tk8.4
    export TCL_LIBRARY TK_LIBRARY
fi

gvx="${GEOMVIEW_GVX}"
case "$gvx" in
  *gvx.[a-zA-Z3])
    gvx="`expr match \"$GEOMVIEW_GVX\" '(.*)\.[^./]*' \| \"$GEOMVIEW_GVX\"`" ;;
esac

can_C3D() {
  case "$DISPLAY" in
    :*|unix:*) test -w /dev/fbs/ffb0 || return 1 ;;
  esac
  case `ldd "${gvx}.C3D" 2>&1` in
    *found*|*"No such"*) return 1 ;;
  esac
}

gvx_option() {
 case "$1" in
  -ogl|-OGL|-opengl|-OpenGL|-OPENGL)
	suf=.OGL; suf2=.MESA
	if [ "$MACHTYPE" = "solaris" ]; then
	    suf=.MESA
	    if can_C3D; then suf=.C3D; fi
	fi
	shift ;;
  -mesa|-Mesa|-MESA)
	suf=.MESA; suf2=.OGL ;;
  -x11|-X11)
	suf=.X11; shift ;;
  -c3d|-C3D|-Creator3D|-creator3d)
	suf=.C3D; suf2=.MESA ;;
  -h|-help|-HELP|-\?*)
    echo "For advice on selecting which geomview graphics option to use,
see also \"geomview -graphics\"."
    exit 0
    ;;
  -grap*|-Grap*|-GRAP*)
    if [ -f "$gvx.X11" -o -f "$gvx.MESA" -o -f "$gvx.C3D" -o -f "$gvx.OGL" ]; then
	g=`basename $0`
	echo "The $0 script can invoke any of several executables:"
	test -f "$gvx.X11" && \
	  echo "  with geomview's own X software rendering: use $g -x11"
	test -f "$gvx.MESA" && \
	  echo "  with Mesa OpenGL-like software rendering: use $g -mesa"
	test -f "$gvx.X11" && \
	  echo "  with Sun OpenGL for UltraSparcs with Creator3D: use $g -creator3d"
	test -f "$gvx.OGL" && \
	  echo "  with Open GL: use $g -opengl"
	echo "If given, this option MUST BE THE FIRST on geomview's command line."
	echo "Using some variant of Open GL allows transparency and texture-mapping;"
	echo "software OpenGL implementations are slower than $g -x11, but hardware ones"
	echo "run faster."
	echo "To change the default, setenv GEOMVIEW_GVXOPT to the appropriate option."
    else
	echo "No alternative graphics options available."
    fi >&2
    exit 0
    ;;
  *) test ;;
 esac
}

gvx_option ${GEOMVIEW_GVXOPT}

if gvx_option "$1"; then shift; fi


if [ "$MACHTYPE" = "solaris" -a -z "$suf" -a -w /dev/fbs/ffb0 ] && can_C3D; then
  echo "$0: Invoking Creator3D version (see \"geomview -graphics\" for info)" >&2
  suf=.C3D
fi

if [ -f "$gvx$suf" ]; then
  GEOMVIEW_GVX="$gvx$suf"
elif [ -f "$gvx$suf2" ]; then
  GEOMVIEW_GVX="$gvx$suf2"
elif [ -f "$gvx" ]; then
  GEOMVIEW_GVX="$gvx"
fi
export GEOMVIEW_GVX

case "$1" in
  -dbx) GEOMVIEW_DEBUG=1 exec dbx $GEOMVIEW_GVX ;;
  -edge) GEOMVIEW_DEBUG=1 exec edge $GEOMVIEW_GVX ;;
  -gdb) GEOMVIEW_DEBUG=1 exec gdb $GEOMVIEW_GVX ;;
  -xdb) GEOMVIEW_DEBUG=1 exec xdb $GEOMVIEW_GVX ;;
  -ddd) GEOMVIEW_DEBUG=1 exec ddd $GEOMVIEW_GVX ;;
  *) exec $GEOMVIEW_GVX ${1+"$@"} ;;
esac
